src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -0,0 +1,167 @@
+
+public class BulkImportUserUtils {
+    public static BulkImportUser createBulkImportUserFromJSON(JsonObject userData, String id) throws InvalidBulkImportDataException {
+        List<String> errors = new ArrayList<>();
+
+        String externalUserId = parseAndValidateField(userData, "externalUserId", ValueType.STRING, false, String.class, errors, ".");
+        JsonObject userMetadata = parseAndValidateField(userData, "userMetadata", ValueType.OBJECT, false, JsonObject.class, errors, ".");
+        List<String>  userRoles = getParsedUserRoles(userData, errors);
+        List<TotpDevice> totpDevices = getParsedTotpDevices(userData, errors);
+        List<LoginMethod> loginMethods = getParsedLoginMethods(userData, errors);
+
+        if (!errors.isEmpty()) {
+            throw new InvalidBulkImportDataException(errors);
+        }
+        return new BulkImportUser(id, externalUserId, userMetadata, userRoles, totpDevices, loginMethods);
+    } 
+
+    private static List<String> getParsedUserRoles(JsonObject userData, List<String> errors) {
+        JsonArray jsonUserRoles = parseAndValidateField(userData, "roles", ValueType.ARRAY_OF_STRING,
+                false,
+                JsonArray.class, errors, ".");
+
+        if (jsonUserRoles == null) {
+            return null;
+        }
+
+        List<String> userRoles = new ArrayList<>();
+        jsonUserRoles.forEach(role -> userRoles.add(role.getAsString().trim()));
+        return userRoles;
+    }
+
+    private static List<TotpDevice> getParsedTotpDevices(JsonObject userData, List<String> errors) {
+        JsonArray jsonTotpDevices = parseAndValidateField(userData, "totp", ValueType.ARRAY_OF_OBJECT, false, JsonArray.class, errors, ".");
+        if (jsonTotpDevices == null) {
+            return null;
+        }
+
+        List<TotpDevice> totpDevices = new ArrayList<>();
+        for (JsonElement jsonTotpDeviceEl : jsonTotpDevices) {
+            JsonObject jsonTotpDevice = jsonTotpDeviceEl.getAsJsonObject();
+
+            String secretKey = parseAndValidateField(jsonTotpDevice, "secretKey", ValueType.STRING, true, String.class, errors, " for a totp device.");
+            Number period = parseAndValidateField(jsonTotpDevice, "period", ValueType.NUMBER, true, Number.class, errors, " for a totp device.");
+            Number skew = parseAndValidateField(jsonTotpDevice, "skew", ValueType.NUMBER, true, Number.class, errors, " for a totp device.");
+            String deviceName = parseAndValidateField(jsonTotpDevice, "deviceName", ValueType.STRING, false, String.class, errors, " for a totp device.");
+
+            if (period != null && period.intValue() < 1) {
+                errors.add("period should be > 0 for a totp device.");
+            }
+            if (skew != null && skew.intValue() < 0) {
+                errors.add("skew should be >= 0 for a totp device.");
+            }
+
+            if(deviceName != null) {
+                deviceName = deviceName.trim();
```

Comment:
don't we make this lowerCase too in the API for creating a device?

Reply:
We don't make deviceName lowercase.